pinot-core/src/main/java/org/apache/pinot/core/query/pruner/SelectionQuerySegmentPruner.java

```diff
@@ -76,13 +76,18 @@ public void init(PinotConfiguration config) {
     }
 
     // If LIMIT is not 0, only prune segments for selection queries without filter
-    // TODO: This does not take upsert valid doc ids filter into account, which can cause less than expected records
-    //       returned for upsert table
     FilterContext filter = queryContext.getFilter();
     if (filter != null) {
       return segmentDataManagers;
     }
 
+    // Do not prune for upsert table
+    if (!segmentDataManagers.isEmpty()) {
+      if (segmentDataManagers.get(0).getSegment().getValidDocIndex() != null) {
```

Comment:
Can valid doc index be null if there are no more valid docs in that first segment? Or, is it guaranteed that this (potential optimization) will never happen

Reply:
For now, the implementation guarantees all segments (immutable+mutable) of upsert table have non-null valid docs, while the segments of append-only table have null values.

But it's a good question that if we will use this valid doc for other purposes. By design, this valid doc is quite general and we may consider other types of ingestion-time filtering.